lms/migrations/versions/52755322151e_db_migration_to_back_fill_organizations.py

```diff
@@ -0,0 +1,194 @@
+def connected_subgraphs(edges):
+    """
+    Get a list of sets of nodes which form connected sub-graphs.
+
+    * All edges are assumed to be bidirectional
+    * All items in a set are guaranteed to be at least indirectly connected
+    * No items in different sets should be connected even indirectly
+    """
+
+    # Put every node in a group on its own
+    group_to_nodes = {node: {node} for node in itertools.chain(*edges)}
+    # ... and record it's location (which will change)
+    node_to_group = {node: node for node in group_to_nodes.keys()}
+
+    # Repeatedly merge groups if they are joined by an edge
+    for left_node, right_node in edges:
+        left_group, right_group = node_to_group[left_node], node_to_group[right_node]
+
+        if left_group == right_group:
+            # Nothing to do, they are already in the same group
+            continue
+
+        # Merge the right group into the left and update the lookups
+        for node in group_to_nodes.pop(right_group):
+            group_to_nodes[left_group].add(node)
+            node_to_group[node] = left_group
+
+    # Discard our internal ids for the sub-graphs as they have no meaning
+    # outside this method. Sort them to make it easier on the eye.
+    return sorted(sorted(list(nodes)) for nodes in group_to_nodes.values())
+
+
+def pick_name(names):
+    """Pick the most common or the longest non-empty name."""
+
+    names = [name.strip() for name in names if name and name.strip()]
+    if not names:
+        return None
+
+    if len(names) == 1:
+        return names[0]
+
+    most_common = Counter()
+    longest_name = ""
+    for name in names:
+        most_common[name] += 1
+        if len(name) > len(longest_name):
+            longest_name = name
+
+    for name, count in most_common.most_common(1):
+        if count > 1:
+            return name
+
+    return longest_name
+
+
+def upgrade():
+    db_session = sa.orm.Session(bind=op.get_bind())
+
+    print("Clearing old organizations...")
+    op.execute("UPDATE application_instances SET organization_id = NULL")
+    op.execute("DELETE FROM organization")
+
+    with open(__file__.replace(".py", ".sql"), encoding="utf-8") as handle:
+        query = handle.read()
+
+    print("Detecting GUIDs...")
+    edges = list(db_session.execute(query))
+
+    print("Grouping GUIDs...")
+    grouped_guids = connected_subgraphs(edges)
+
+    print("Creating new organizations...")
+    total = 0
+
+    for ai_ids in grouped_guids:
+        ais = (
+            db_session.query(ApplicationInstance)
+            .filter(ApplicationInstance.id.in_(ai_ids))
+            .all()
+        )
+        if not ais:
+            continue
+
+        organization = Organization(
+            name=pick_name([ai.tool_consumer_instance_name for ai in ais])
+        )
+        total += 1
+        print(
+            f"\tCreating '{organization.name}' for {len(ais)} application instance(s)"
+        )
```

Comment:
This looks like it might print out a lot, but it should only create about 1200 max given the number of GUIDs we have